### PR TITLE
v4 API

### LINKED ISSUE #52: Too easy to create infinite loops
Due to React using shallow reference equality, it's a bit too easy to cause infinite loops by putting something like a Maybe in your component state and rebuilding it in an unconditional `setState` inside `receiveProps`.

Possible fixes:
- just revert the 2.0.1 change, `findDOMNode` will be slightly risky if not used within a `setStateThen` callback but this is a bit of an edge case anyway
- expire the slightly overloaded `receiveProps`, using optional lifecycle hooks instead (bigger API change)
   - [reason-react](https://reasonml.github.io/reason-react/docs/en/state-actions-reducer)'s approach might be worth considering
- something else I'm not thinking of?

### LINKED ISSUE #53: Performance and potential changes to allow typeclasses on props
As our applications grow, the disconnect between PureScript's tendency toward constantly creating functions (currying, lots of callback props) and objects (data constructors) and React's reliance on reference equality for optimization is becoming apparent. There are a few cases where we'd like to use Eq or a manual equality implementation rather than React's shallow reference equality checks, but this isn't possible with the current API.

Additionally, the new context API is a lot more type-friendly so this might be a good time to experiment with that a little.

After exploring reason-react's approach a bit, I think both typeclasses and context may be possible by splitting the "createClass" and "implement render and lifecycles" bits into separate pieces:

```purs
-- | technically this isn't pure, but that's already the
-- | case for react-basic 2's `component` function
componentClass :: forall props. React.Class props
componentClass = component "Component"

-- | currently `Component` is synonymous with JS React
-- | classes, but it might make sense to change that and
-- | add a `toReactClass` utility instead (you'd need to resolve
-- | any typeclass constraints when calling `toReactClass`)
component :: forall props. Eq props => React.Component props
component = ...combine render and lifecycles with `componentClass` (reason-react uses object spread, maybe a helper function would be necessary)
  where
    render props = ...
```

Having `componentClass` extracted from the main component definition keeps it from being recreated on every render (this currently prevents typeclass use). It also opens the api up a bit for distinguishing "configuration" from "props" (or maybe "static props" from "dynamic props"). The former would be arguments which don't change as your application runs, while the latter would be passed as normal React props. This means api implementations and callbacks can be passed without tricking React into always rerendering:

```purs
x config = stateless { render }
  where
    render props =
      element someComponent { getData, onClick } { id }
```

This means less work for React itself, but it also allows this optimization:

```purs
x config = stateless { render }
  where
    render props =
      someComponentPreapplied { id }

    someComponentPreapplied = element someComponent { getData, onClick }
```

### PR DESCRIPTION
**Pull request by spicydonuts**
These changes address a few things:

- Allow typeclasses in component props or state
- Allow any types as props and state, not just records
- Support lifecycle functions such as `shouldUpdate` and `willUnmount`
- Encourage (but not require) an Action-like API for component state

I think this is ready enough for feedback. The API is extremely `reason-react` inspired (Reason has similar language requirements/constraints, the API is very close to vanilla React's besides the concept of Actions/update/reducer because Reason and PureScript have more powerful pattern matching abilities).

There are still a few bits I'm not super happy with -- I'll annotate the code to point them out.

This is a pretty significant change over `react-basic` 2.0. It's now less "basic" in the "implement a small subset of vanilla React in PS, where 'basic' means small" and instead more along the lines of "implement a thin but opinionated wrapper around React, where 'basic' means encourage correctness without sacrificing performance or type complexity". :slightly_smiling_face: 

Remaining work:

- [x] Review API, feedback, etc
- [x] Write documentation
- [x] Update the Lumi component library to use this branch to see how it feels and compare performance

Fixes #52
Fixes #53

### CODE REVIEW

File: examples/component/src/Main.purs [hunk 1]
```diff
@@ -19,5 +18,5 @@ main = do
   case container of
     Nothing -> throw "Container element not found."
     Just c  ->
-      let app = element Container.component {}
+      let app = Container.render
```
[the remark this answers is not in the document]
Comment (spicydonuts) [reply examples/component/src/Main.purs]: I added `toReactComponent`. It ends up just being an unsafe coerce, but it’s still an important distinction because ReactComponent reference equality matters while `props -> JSX` reference equality does not (meaning no type class wrappers)

**Comment by spicydonuts** [comment]
@justinwoo @paluh @zudov Not sure which of you are actively using this library, but we're curious what your thoughts are on these changes if you are using it. (I haven't written a readme update yet, but there is a `React.Basic.Compat` module as well for delaying the update of particular components.

**Comment by justinwoo** [comment]
I'm starting to use more of my own layer on top that just works via a version of the regular React component API, so I won't have much to say about this immediately. Hopefully I will get around to reading through this sometime.

**Comment by spicydonuts** [comment]
Ah, @f-f too, thanks!

**Comment by f-f** [comment]
@spicydonuts super nice, thanks for the ping! I read through this a couple of times by now, and it looks really good. I especially like the `Action`-like API and the fact that we don't have to do things like `preventDefault >>> ..` by hand 👍 

For this week's meetup I [implemented TodoMVC with `react-basic`](https://github.com/f-f/purescript-react-basic-todomvc), I could test-drive this new API with it

/cc @vapaj

**Comment by spicydonuts** [comment]
Neat! I'd love to see how it goes (and hopefully if I push a couple more tweaks it won't be too confusing -- let me know if you have questions). Also, while you _can_ port v2/3 to this mechanically, it feels a lot better if you give the actions are thought out a bit (send/capture work better, update is more descriptive, etc).

**Comment by spicydonuts** [comment]
There is currently a compiler issue/limitation which shows up occasionally when using `capture` in a `where` function. You can work around it by annotating the function (for render: `Self Props State Action -> JSX`). It's one of the remaining things I want to fix.

File: src/React/Basic.purs [hunk 2]
```diff
@@ -27,115 +35,184 @@ instance semigroupJSX :: Semigroup JSX where
...
+type ComponentSpec props state action =
+  { "$$type" :: ReactComponent props
+  , initialState :: state
+  , shouldUpdate :: LimitedSelf props state -> props -> state -> Boolean
+  , didMount :: Self props state action -> Effect Unit
+  , didUpdate :: Self props state action -> Effect Unit
+  , willUnmount :: LimitedSelf props state -> Effect Unit
+  , update :: Update props state action
+  , render :: Self props state action -> JSX
+  }
```
[the remark this answers is not in the document]
Comment (zudov) [reply src/React/Basic.purs]: It would be handy to have a couple of functions that allow changing the component type.

In our case we want to use the component both from JS and Purescript, we have this helper:

```haskell
contramapComponentProps
  :: forall state propsA propsB
   . (propsB -> propsA)
  -> ComponentSpec propsA state
  -> ComponentSpec propsB state
contramapComponentProps mapProps spec = spec
  { receiveProps = \self -> spec.receiveProps $ self
      { props = mapProps self.props }
  , render = \self -> spec.render $ self
      { props = mapProps self.props }
  }
```

which allows us to define our components with a pattern like:

```haskell
type Props =
  { foo :: Maybe Bar }

type JSProps =
  { foo :: Nullable Bar } 

fromJSProps :: JSProps -> Props

componentSpec :: ComponentSpec Props State

component :: Component Props
component = React.component componentSpec

jsComponent :: Component JSProps
jsComponent = React.component $ React.contramapComponentProps fromJSProps componentSpec
```

We haven't yet had a need for changing the `state` type (since state is internal), but I guess it might come useful (as well as changing the action type).

[the remark this answers is not in the document]
Comment (spicydonuts) [reply src/React/Basic.purs]: That makes sense. In this PR, `Component` is an alias for `ComponentSpec` the `react-basic` specification of a component, not an actual React component. Because of that, this PR adds the `ReactComponent props` type and `toReactComponent :: ComponentSpec props _ _ -> ReactComponent`. Maybe a version of `toReactComponent` could provide this prop mapping you're doing.

[the remark this answers is not in the document]
Comment (spicydonuts) [reply src/React/Basic.purs]: We could just change the type of `toReactComponent`, since `identity` is easy enough to pass in:
```purs
toReactComponent
  :: forall props state action jsProps
   . ({ | jsProps } -> props)
  -> ComponentSpec props state state action
  -> ReactComponent { | jsProps }
```

[the remark this answers is not in the document]
Comment (spicydonuts) [reply src/React/Basic.purs]: This exists now

**Comment by spicydonuts** [comment]
Refactored `Self` a bit to avoid the unexpected type error sometimes seen with `capture`. API's nearly the same, but a few functions which were found on `Self` before are now just normal, module-level functions with `Self` as a parameter. These changes allowed for some minor performance improvements as well :)

File: src/React/Basic.purs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zudov) [reply src/React/Basic.purs]: @spicydonuts This is a lovely solution, thanks.

**Comment by jgoux** [comment]
Now that React disclosed hooks (https://reactjs.org/docs/hooks-reference.html), what do you think about them and how could they be integrated with PureScript?

**Comment by spicydonuts** [comment]
They're somewhat monadic which is interesting. The implementation and usage in JS is concerning, but that's already the case for a lot of JS anyway. Maybe it could be exposed via PureScript in a safer way, but whether it's a nicer API would take some time and experimentation to discover. It's also just a proposal at this time, so it's not ready to build on top of anyway.

tl;dr, not sure yet

**Comment by jgoux** [comment]
Thanks for your answer!

The experimentation has begun with Reason : https://twitter.com/jaredforsyth/status/1056902656197828608 🎊

**Comment by spicydonuts** [comment]
@jgoux So, I took a stab at it: [Counter example with hooks](https://github.com/spicydonuts/purescript-react-basic/blob/hooks/examples/counter/src/Counter.purs)

I actually am liking it.. :thinking: 

I saw that alt design too, and the interesting bit is that it's no different in PS.. do-syntax reads like the React proposal but compiles to the alt proposal :wink:

**Comment by spicydonuts** [comment]
It feels much more flexible and opt-in (to state and lifecycles), and makes it easier to build complicated state helpers that manage their own subscriptions outside the component.

**Comment by jgoux** [comment]
Wow this API seems awesome!
I think I might try using it on a small project !

**Comment by spicydonuts** [comment]
I removed `Applicative Render` -- that inner `props -> do ...` being passed to `component` is `props -> Render JSX`. `Render` is a newtype around Effect with its constructor hidden, and all the hooks run in `Render`. Removing `pure` makes it harder to do conditional logic, but it's still possible.. so maybe it isn't worth the inconvenience of not having `pure`.. I can add it back

**Comment by spicydonuts** [comment]
Actually, yeah.. it'd make it significantly harder to write custom state/effect logic without `pure` undoing that haha

**Comment by spicydonuts** [comment]
Made a separate PR to discuss hooks: https://github.com/spicydonuts/purescript-react-basic/pull/1

File: examples/counter/src/Counter.purs [hunk 3]
```diff
@@ -2,31 +2,31 @@ module Counter where
...
+data Action
+  = Increment
+
+counter :: Props -> JSX
+counter = make component { initialState, update, render }
   where
-    initialState =
-      { counter: 0
-      }
+    initialState = { counter: 0, dummy: 0 }
```
Comment (paf31) [inline examples/counter/src/Counter.purs]: `dummy`?

Comment (spicydonuts) [reply examples/counter/src/Counter.purs]: Oops, will remove
